<a href="https://colab.research.google.com/github/2403a52038-star/aiac/blob/main/Lab3_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Task 1:Try 3 different prompts to generate a factorial function

# Task
Generate a Python function to calculate the factorial of a number using 3 different prompts.

## Generate code with prompt 1

### Subtask:
Generate a Python function to calculate the factorial of a number using the first prompt.


**Reasoning**:
Generate a Python function to calculate the factorial of a non-negative integer using the first prompt.



In [38]:
# Prompt 1: Generate a Python function that calculates the factorial of a non-negative integer.

def factorial(n):
  """Calculates the factorial of a non-negative integer.

  Args:
    n: A non-negative integer.

  Returns:
    The factorial of n.

  Raises:
    ValueError: If n is a negative integer.
  """
  if not isinstance(n, int) or n < 0:
    raise ValueError("Input must be a non-negative integer")
  if n == 0:
    return 1
  else:
    return n * factorial(n-1)


## Generate code with prompt 2

### Subtask:
Generate a Python function to calculate the factorial of a number using the second prompt.


**Reasoning**:
Add a new code cell with the specified comment and the Python function for calculating factorial using an iterative approach based on the second prompt.



In [39]:
# Prompt 2: Generate a Python function that calculates the factorial of a non-negative integer using an iterative approach.

def factorial_iterative(n):
  """Calculates the factorial of a non-negative integer using an iterative approach.

  Args:
    n: A non-negative integer.

  Returns:
    The factorial of n.

  Raises:
    ValueError: If n is a negative integer.
  """
  if not isinstance(n, int) or n < 0:
    raise ValueError("Input must be a non-negative integer")
  if n == 0:
    return 1
  else:
    result = 1
    for i in range(1, n + 1):
      result *= i
    return result

## Generate code with prompt 3

### Subtask:
Generate a Python function to calculate the factorial of a number using the third prompt.


**Reasoning**:
Write a prompt that uses a mathematical library and generate the Python function using that prompt.



In [40]:
# Prompt 3: Generate a Python function that calculates the factorial of a non-negative integer using the math module.

import math

def factorial_math(n):
  """Calculates the factorial of a non-negative integer using the math module.

  Args:
    n: A non-negative integer.

  Returns:
    The factorial of n.

  Raises:
    ValueError: If n is a negative integer.
  """
  if not isinstance(n, int) or n < 0:
    raise ValueError("Input must be a non-negative integer")
  return math.factorial(n)


## Test the functions

### Subtask:
Test each generated function with sample inputs to ensure they work correctly.


**Reasoning**:
Define sample inputs and test each factorial function with these inputs, printing the results for comparison.



In [41]:
# 1. Define a list of sample non-negative integer inputs.
sample_inputs = [0, 5, 10]

# 2. and 3. Test each function and print the results.
print("Testing factorial function:")
for n in sample_inputs:
  result = factorial(n)
  print(f"  factorial({n}) = {result}")

print("\nTesting factorial_iterative function:")
for n in sample_inputs:
  result = factorial_iterative(n)
  print(f"  factorial_iterative({n}) = {result}")

print("\nTesting factorial_math function:")
for n in sample_inputs:
  result = factorial_math(n)
  print(f"  factorial_math({n}) = {result}")

# 4. Compare the outputs (implicitly done by printing side-by-side for each input).

Testing factorial function:
  factorial(0) = 1
  factorial(5) = 120
  factorial(10) = 3628800

Testing factorial_iterative function:
  factorial_iterative(0) = 1
  factorial_iterative(5) = 120
  factorial_iterative(10) = 3628800

Testing factorial_math function:
  factorial_math(0) = 1
  factorial_math(5) = 120
  factorial_math(10) = 3628800


## Summary:

### Data Analysis Key Findings

*   Three different Python functions were successfully generated to calculate the factorial of a non-negative integer:
    *   `factorial`: A recursive implementation.
    *   `factorial_iterative`: An iterative implementation.
    *   `factorial_math`: An implementation utilizing the `math.factorial` function from the `math` module.
*   Each generated function includes input validation to ensure the input is a non-negative integer and handles the base case where the factorial of 0 is 1.
*   Testing with sample inputs (0, 5, and 10) showed that all three functions produced the correct factorial values:
    *   Factorial of 0: 1
    *   Factorial of 5: 120
    *   Factorial of 10: 3,628,800
*   All three implementations are functionally equivalent for calculating the factorial of non-negative integers.

### Insights or Next Steps

*   The exercise demonstrates how different prompts can lead to variations in code implementation (recursive, iterative, built-in function).
*   For practical applications requiring factorial calculations, using `math.factorial` is generally the most efficient and concise approach in Python.


Task 2:Provide a clear example input-output prompt to generate a sorting function

In [42]:
def sort_list(input_list):
  """Sorts a list of numbers in ascending order.

  Args:
    input_list: A list of numbers.

  Returns:
    A new list containing the numbers sorted in ascending order.
  """
  return sorted(input_list)

# Example usage based on the prompt:
my_list = [3, 1, 4, 1, 5, 9, 2, 6]
sorted_list = sort_list(my_list)
print(sorted_list)

[1, 1, 2, 3, 4, 5, 6, 9]


Here's an example of an input-output prompt to generate a sorting function:

Input: Given a list of numbers [3, 1, 4, 1, 5, 9, 2, 6], write a Python function sort_list that sorts the list in ascending order.

Expected Output: The sorted list should be [1, 1, 2, 3, 4, 5, 6, 9].

Task 3:Start with the vague prompt “Generate python code to calculate power bill” and
improve it step-by-step

In [44]:
# Generate python code to calculate power bill given the electricity consumption in kWh and the cost per kWh.

def calculate_power_bill(kwh_consumption, cost_per_kwh):
  """Calculates the power bill based on consumption and rate.

  Args:
    kwh_consumption: The amount of electricity consumed in kWh.
    cost_per_kwh: The cost per kWh.

  Returns:
    The calculated power bill.
  """
  return kwh_consumption * cost_per_kwh

# Example usage:
consumption = 150  # kWh
rate = 0.12      # $/kWh
bill = calculate_power_bill(consumption, rate)
print(f"The power bill is: ${bill:.2f}")

The power bill is: $18.00


This code defines a Python function called calculate_power_bill that takes two arguments: kwh_consumption (the amount of electricity used) and cost_per_kwh (the price per unit of electricity). It calculates the total bill by multiplying these two values and then prints an example of how to use the function.

Task 4:Write structured comments to help AI generate two linked functions (e.g., login_user()
and register_user())

In [46]:
# Demonstrate shared logic: Accessing the user_database

# After registering a user
print("\nUser database after registration:")
print(user_database)

# After a successful login (no change to database, but demonstrates access)
print("\nUser database after successful login attempt:")
# The database itself doesn't change on login, but the function accesses it to verify
print(user_database)

# After a failed login attempt (no change to database, but demonstrates access)
print("\nUser database after failed login attempt:")
# The database itself doesn't change on login, but the function accesses it to verify
print(user_database)


User database after registration:
{'test_user': 'password123_hashed'}

User database after successful login attempt:
{'test_user': 'password123_hashed'}

User database after failed login attempt:
{'test_user': 'password123_hashed'}


This code snippet prints the current state of the user_database dictionary at different points after calling the register_user and login_user functions. It demonstrates that both functions access and potentially modify the user_database, which is their shared logic. It shows the database content after a registration and then after successful and failed login attempts to illustrate how the functions interact with the stored user data

Task 5:Analyzing Prompt Specificity: Improving Temperature Conversion Function with
Clear Instructions

In [54]:
# Generate python code for temperature conversion (based on vague prompt)

def convert_temperature(temp, from_unit, to_unit):
  """Converts temperature between units (basic example).

  Args:
    temp: The temperature value.
    from_unit: The unit to convert from (e.g., 'C', 'F').
    to_unit: The unit to convert to (e.g., 'C', 'F').

  Returns:
    The converted temperature, or None if conversion is not supported.
  """
  if from_unit == 'C' and to_unit == 'F':
    return (temp * 9/5) + 32
  elif from_unit == 'F' and to_unit == 'C':
    return (temp - 32) * 5/9
  # Add more conversions as needed
  else:
    return None

# Example usage (assuming Celsius to Fahrenheit)
celsius_temp = 25
fahrenheit_temp = convert_temperature(celsius_temp, 'C', 'F')
print(f"{celsius_temp}°C is equal to {fahrenheit_temp}°F")

25°C is equal to 77.0°F


This code defines a simple Python function convert_temperature that takes a temperature value, the original unit, and the target unit as input. It currently supports converting Celsius to Fahrenheit and Fahrenheit to Celsius. The example shows how to convert 25°C to Fahrenheit.